Implement `jsctl clusters cleanup secrets remove-secret-owner-refs`

charlieegan3 opened the pull request:
Implement `jsctl clusters cleanup secrets remove-secret-owner-refs`. This command is to help users prepare their clusters to back up, remove and replace cert-manager CRDs safely.

https://user-images.githubusercontent.com/1774239/201705713-c52b358b-5eb5-4509-920e-fd69f6caf65c.mov

irbekrm commented on internal/command/clusters/cleanup.go:
```diff
@@ -78,6 +81,68 @@ func removeSecretOwnerReferences(run types.RunFunc, kubeConfigPath string) *cobr
+						return fmt.Errorf("error marshalling new secret: %s", err)
+					}
+
+					patch, err := strategicpatch.CreateTwoWayMergePatch(secretData, newSecretData, v1core.Secret{})
```
I think this might not work as expected- I tested with a Secret that had an extra owner reference and saw that being removed too. I think that we probably want to use [SSA](https://kubernetes.io/docs/reference/using-api/server-side-apply/) here as the strategicpatch [seems deprecated](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-api-machinery/strategic-merge-patch.md)

irbekrm replied:
I think it is possible to take over ownership with SSA
however, I see that the latest changes fix the problem 👍🏼

irbekrm reviewed (changes requested):
Thank you for the good work on this @charlieegan3 , I think the refactor also makes sense and makes the codebase cleaner.

I've tried to be thorough with reviewing this PR as if this command is actually used by folks then it will be in production environments and the effect of it failing could be quite bad.

I've tested a number of combinations and left a bunch of comments.

Additionally:
- I think the command should be 'experimental'
- It would be better if it could output the resources that will be changed and ask user for approval (i.e like `terraform apply`)
- It would be good if the command stated the number of resources that got modified (at the moment nothing shows if there were no resources to be modified)
- at the moment this seems to remove unrelated owner refs as well as those of certificates
- I wonder if we shouldn't just have the `clusters cleanup secrets` do the owner ref removal and get rid of the `remove-secret-owner-ref` subcommand. We had a discussion about this that I don't fully remember

irbekrm commented:
Also noticed that the command errors out if `--enable-certificate-owner-ref` flag is set to false

charlieegan3 commented on internal/command/operator.go:
```diff
@@ -12,6 +12,7 @@ import (
 	"gopkg.in/yaml.v2"
 
 	"github.com/jetstack/jsctl/internal/client"
+	errors2 "github.com/jetstack/jsctl/internal/command/errors"
```
Thanks, addressed in d43037dcac5e7de8bcdeb6af2074704ed550616f

charlieegan3 commented on internal/command/clusters/cleanup.go:
```diff
@@ -39,7 +44,42 @@ func removeSecretOwnerReferences(run types.RunFunc, kubeConfigPath string) *cobr
+			if found, _ := certManagerStatus.GetControllerFlagValue(enableCertificateOwnerRefFlag); found {
+				fmt.Fprintf(os.Stderr, "cert-manager controller has --%s flag set, this must be removed from the controller's deployment pod spec before continuing.\n\n", enableCertificateOwnerRefFlag)
+				fmt.Fprintf(os.Stderr, "If left set, the controller will re-add owner references to secrets, which will cause them to be deleted as part of the backup process when Certificates are removed from the cluster state. This will lead to reissueance of certificates on restore and could cause potential outages caused by CA rate limiting.\n\n")
+				fmt.Fprintf(os.Stderr, "Note: this might require an update to configuration management for the cluster, GitOps-based (e.g. Flux) or otherwise (Jenkins etc).\n")
```
Fixed in 071863801c08977b3c87de98213197587b7b3d80

charlieegan3 commented on internal/command/clusters/cleanup.go:
```diff
@@ -39,7 +44,42 @@ func removeSecretOwnerReferences(run types.RunFunc, kubeConfigPath string) *cobr
+
+			enableCertificateOwnerRefFlag := "enable-certificate-owner-ref"
+			if found, _ := certManagerStatus.GetControllerFlagValue(enableCertificateOwnerRefFlag); found {
+				fmt.Fprintf(os.Stderr, "cert-manager controller has --%s flag set, this must be removed from the controller's deployment pod spec before continuing.\n\n", enableCertificateOwnerRefFlag)
```
Fixed in 997ba1d89e2fb996ba6ccc0b268c9d0ccf93aa4f

charlieegan3 commented on internal/command/clusters/cleanup.go:
```diff
@@ -39,7 +44,42 @@ func removeSecretOwnerReferences(run types.RunFunc, kubeConfigPath string) *cobr
+			enableCertificateOwnerRefFlag := "enable-certificate-owner-ref"
+			if found, _ := certManagerStatus.GetControllerFlagValue(enableCertificateOwnerRefFlag); found {
+				fmt.Fprintf(os.Stderr, "cert-manager controller has --%s flag set, this must be removed from the controller's deployment pod spec before continuing.\n\n", enableCertificateOwnerRefFlag)
+				fmt.Fprintf(os.Stderr, "If left set, the controller will re-add owner references to secrets, which will cause them to be deleted as part of the backup process when Certificates are removed from the cluster state. This will lead to reissueance of certificates on restore and could cause potential outages caused by CA rate limiting.\n\n")
```
Fixed in 9397f2c2c9e6b8bed0a1a26726bcb982255c44bb

charlieegan3 commented on internal/command/clusters/cleanup.go:
```diff
@@ -2,11 +2,14 @@ package clusters
 
 	"github.com/spf13/cobra"
 	v1core "k8s.io/api/core/v1"
+	v1meta "k8s.io/apimachinery/pkg/apis/meta/v1"
```
fixed as part of 665df89

charlieegan3 commented on internal/command/clusters/cleanup.go:
```diff
@@ -78,6 +81,68 @@ func removeSecretOwnerReferences(run types.RunFunc, kubeConfigPath string) *cobr
+				return nil
+			}
+
+			fmt.Fprintf(os.Stderr, "Checking for ownerReferences on secrets...\n")
```
Fixed in c7b2efa8b42d9d6f2144ad88f260081e806a7af3

charlieegan3 commented on internal/command/clusters/cleanup.go:
```diff
@@ -78,6 +81,68 @@ func removeSecretOwnerReferences(run types.RunFunc, kubeConfigPath string) *cobr
+
+			// if the flag is not found, we still want to check that the owner
+			// references are not present. It can take some time for
+			// cert-manager to remove them, even if cert-manager is running.
```
Updated in d5148d6b1bcc7589e14a693e499a681a109454d4

charlieegan3 commented on internal/command/clusters/cleanup.go:
```diff
@@ -0,0 +1,46 @@
+	return &cobra.Command{
+		Use:   "remove-secret-owner-refs",
+		Short: "Remove owner references to cert-manager resources from all secrets in the cluster",
+		Long:  "Removing owner references from secrets allows cert-manager CRDs to be removed and the underlying secret data to be retained.",
```
Thanks, fixed in af24d36d4a2d2bb13aba75acdacf05715d5eb073

charlieegan3 commented on internal/command/clusters/cleanup.go:
```diff
@@ -0,0 +1,46 @@
+func removeSecretOwnerReferences(run types.RunFunc, kubeConfigPath string) *cobra.Command {
+	return &cobra.Command{
+		Use:   "remove-secret-owner-refs",
+		Short: "Remove owner references to cert-manager resources from all secrets in the cluster",
```
Fixed in af24d36d4a2d2bb13aba75acdacf05715d5eb073

charlieegan3 commented on internal/command/clusters/cleanup.go:
```diff
@@ -78,6 +81,68 @@ func removeSecretOwnerReferences(run types.RunFunc, kubeConfigPath string) *cobr
 				fmt.Fprintf(os.Stderr, "If left set, the controller will re-add owner references to secrets, which will cause them to be deleted as part of the backup process when Certificates are removed from the cluster state. This will lead to reissueance of certificates on restore and could cause potential outages caused by CA rate limiting.\n\n")
 				fmt.Fprintf(os.Stderr, "Note: this might require an update to configuration management for the cluster, GitOps-based (e.g. Flux) or otherwise (Jenkins etc).\n")
 				fmt.Fprintf(os.Stderr, "No cleanup action has been taken at this time\n")
```
I have added more help text in 008d65a8234b809638bd7c9deb9aa3cad7969e69

charlieegan3 commented on internal/command/clusters/cleanup.go:
```diff
@@ -0,0 +1,46 @@
+func CleanUp(run types.RunFunc, kubeConfigPath string) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "cleanup",
+		Short: "Perform cleanup operations on a cluster's Kubernetes state",
```
Added in 82a720f9dd5808efe4a331b4204913073dc9f3ea

charlieegan3 commented on internal/command/clusters/cleanup.go:
```diff
@@ -0,0 +1,46 @@
+func secrets(run types.RunFunc, kubeConfigPath string) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "secrets",
+		Short: "Perform cleanup operations related to Kubernetes secrets",
```
Added in c9f6ea7fdbc4d8f19db4f7bf4981e91785d5f481

charlieegan3 commented:
> Also noticed that the command errors out if --enable-certificate-owner-ref flag is set to false

By this, do you mean that the command interprets it as true (which is what I see and need to fix) or do you see another error?

charlieegan3 commented on internal/command/clusters/cleanup.go:
```diff
@@ -39,7 +44,42 @@ func removeSecretOwnerReferences(run types.RunFunc, kubeConfigPath string) *cobr
+			found, err := certManagerStatus.Match(&md)
+
+			if !found {
+				fmt.Fprintf(os.Stderr, "cert-manager not found, nothing to do\n")
```
```
$ go run main.go experimental clusters cleanup secrets remove-certificate-owner-refs
This cluster does not contain any cert-manager Certificates. No action is required.
```

Is this what you intended? 38054ae14f94e76f597e3d89982964470737b472


I think that it makes sense to still check the controller flag though no?

irbekrm replied:
What I meant is that the case where there is no cert-manager deployment, but are CRDs and Certificates is possible (especially as a result of half-done uninstall or something) so it'd make sense to not exit here https://github.com/jetstack/jsctl/commit/38054ae14f94e76f597e3d89982964470737b472#diff-1591e1c83089ec5125d707661e28fe44fcecd460adea1d5fadfac97bd200e300R99

irbekrm replied:
(What you did in https://github.com/jetstack/jsctl/commit/38054ae14f94e76f597e3d89982964470737b472 makes sense, but I think we should also remove the exit if no pod found)

charlieegan3 replied:
Ahh, I see. Sorry, missed that. ce15924b7afdb48cd5fea0f9cf1e2a1bb756228f proceeds even if CM isn't running (while still checking the flag if cert-manager is running)

charlieegan3 commented:
I've updated the command to prompt before taking action:

```
$ go run main.go experimental clusters cleanup secrets remove-certificate-owner-refs
Checking for ownerReferences on secrets containing the issued certificates...
jetstack-secure/exmaple-com-tls needs update
Found 1 secrets with ownerReferences to Certificate resources
Would you like to update the owner references of 1 secrets?
> yes
jetstack-secure/exmaple-com-tls updated
```

I have also addressed the issue with the empty ownerRefs.

Ready for you to have another look now @irbekrm  🙏

irbekrm commented:
Thanks @charlieegan3 , I've ran a couple tests and all looks good to me, I saw that another actor's added owner ref does not get removed this time and that setting the `--enable-certificate-owner-ref` flag to false does not make the command error

The only thing left should be https://github.com/jetstack/jsctl/pull/62#discussion_r1023918822

charlieegan3 commented:
Thanks Irbe, I think that this should be all good now! I have rebased but ce15924b7afdb48cd5fea0f9cf1e2a1bb756228f is the only new commit.

irbekrm reviewed (approved):
Thanks for the good work on this Charlie, I think the command will be useful and the code looks great and should be easy to iterate on in the future!

/lgtm

charlieegan3 commented:
Thanks for all the input, much appreciated 🙂
